doc/tutorials/starship.rst

```diff
@@ -89,12 +102,94 @@ To execute this file using Python, you can use a code block like this:
 
 .. code-block:: console
 
    starship --input test.xbb --output out.txt
 
-After executing the above command, the result will be stored in ``out.txt`` in the current working directory. You can also omit the ``--output`` parameter to print the result to the screen.
+After executing the above command, the result will be stored in ``out.txt`` in the current working directory.
+You can also omit the ``--output`` parameter to print the result to the screen.
+
+
+Program compilation
+===================
+
+In addition to using the program template above, which directly matches the physical
+layout of the hardware device, you can define your interferometer using any combination
+of beamsplitters (:class:`~.ops.BSgate`), rotations/phase shifts (:class:`~.ops.Rgate`).
+Additionally, you can use the :class:`~.ops.Interferometer` command to directly pass a
+unitary matrix to be decomposed and compiled to match the device architecture.
+
+For example, consider the following Blackbird script:
+
+
+.. code-block:: python
+
+   name compilation_example  # Name of the program
+   version 1.0               # Blackbird version number
+   target chip0 (shots=50)   # This program will run on chip0 for 50 shots
+
+   # Define a unitary matrix
+   complex array U[2, 2] =
+      -0.1955885-0.16833594j, 0.77074506+0.58254631j
+      -0.03596574+0.96546083j, 0.00676031+0.2579654j
+
+   # Initial states are two-mode squeezed states,
+   # applied to alternating pairs of modes.
+   S2gate(1.0, 0.0) | [0, 2]
+   S2gate(1.0, 0.0) | [1, 3]
+
+   # Apply the unitary matrix above to
+   # the first pair of modes, as well
+   # as a beamsplitter
+   Interferometer(U) | [0, 1]
+   BSgate(0.543, -0.123) | [0, 1]
+
+   # Duplicate the above unitary for
+   # the second pair of modes
+   Interferometer(U) | [2, 3]
+   BSgate(0.543, -0.123) | [2, 3]
+
+   # Perform a PNR measurement in the Fock basis
+   MeasureFock() | [0, 1, 2, 3]
+
+
+.. note:: You may use :func:`~.random_interferometer` to generate arbitrary random unitaries.
+
+This program will execute following the same steps as above; ``StarshipEngine`` will automatically
+compile the program to match the layout of the chip described in :ref:`first_program`.
+
+You may wish to view the compiled program; this can be easily done in Python using
+the :meth:`~.Program.compile` method:
+
+
+>>> from strawberryfields import StarshipEngine
+>>> from strawberryfields.io import load
+>>> prog = load("test.xbb")
+>>> prog = prog.compile("chip0")
+>>> prog.print()
+S2gate(1, 0) | (q[0], q[2])
+S2gate(1, 0) | (q[1], q[3])
+Rgate(0.9355) | (q[0])
+BSgate(0.7854, 1.571) | (q[0], q[1])
+Rgate(4.886) | (q[0])
+BSgate(0.7854, 1.571) | (q[0], q[1])
+Rgate(-0.3742) | (q[0])
+Rgate(-0.05099) | (q[1])
+Rgate(0.9355) | (q[2])
+BSgate(0.7854, 1.571) | (q[2], q[3])
+Rgate(4.886) | (q[2])
+BSgate(0.7854, 1.571) | (q[2], q[3])
+Rgate(-0.3742) | (q[2])
+Rgate(-0.05099) | (q[3])
+MeasureFock | (q[0], q[1], q[2], q[3])
```

Comment:
:thinking: It doesn't seem to match exactly the version above. There are, e.g., too many `Rgate`'s. Is this ok?

Reply:
Ah good point, the one above is outdated and should match this one.

Reply:
Great, let me know if you need a hand with that bit. Otherwise looks ready to go for me!

Reply:
Just pushed a commit to update the first Blackbird codeblock